### DISCUSSION TITLE
Open file handle issue

### DISCUSSION THREAD

**Discussion by DeanStra** [py-pdf/pypdf#1473] [Q&A]
Hi guys - newbie coder here, sorry for the dumb questions. I was running PyPDF2 to merge a bunch of PDFs

```python
import PyPDF2, os

pdfFiles = []
for filename in os.listdir("."):
    if filename.endswith(".pdf"):
        pdfFiles.append(filename)
pdfFiles.sort()

pdfWriter = PyPDF2.PdfFileWriter()

for filename in pdfFiles:
    pdfFileObj = open(filename, "rb")
    pdfReader = PyPDF2.PdfFileReader(pdfFileObj)

    for pageNum in range(0, pdfReader.numPages):
        pageObj = pdfReader.getPage(pageNum)
        pdfWriter.addPage(pageObj)

pdfOutput = open("allminutes.pdf", "wb")
pdfWriter.write(pdfOutput)
pdfOutput.close()
```
Which is code I got from elsewhere on the internet. Good news - it worked, and is extremely fast!

Bad news - I then tried to move all the pdf files to a different directory, and my pc told me I couldn't, since they were being used by python. This was very annoying, and difficult to resolve.

Bad news - Now when I run the same code, I get the following error:

> ---------------------------------------------------------------------------
> OSError                                   Traceback (most recent call last)
> ~\AppData\Local\Temp/ipykernel_11152/3446078836.py in <module>
>      17 for filename in pdfFiles:
>      18     pdfFileObj = open(filename, 'rb')
> ---> 19     pdfReader = PyPDF2.PdfFileReader(pdfFileObj)
>      20 
>      21     # Loop through all the pages (except the first) and add them.
> 
> ~\anaconda3\lib\site-packages\PyPDF2\_reader.py in __init__(self, stream, strict, warndest, overwriteWarnings)
>     237             with open(stream, "rb") as fileobj:
>     238                 stream = BytesIO(b_(fileobj.read()))
> --> 239         self.read(stream)
>     240         self.stream = stream
>     241 
> 
> ~\anaconda3\lib\site-packages\PyPDF2\_reader.py in read(self, stream)
>     909     def read(self, stream):
>     910         # start at the end:
> --> 911         stream.seek(-1, 2)
>     912         if not stream.tell():
>     913             raise PdfReadError("Cannot read an empty file")
> 
> OSError: [Errno 22] Invalid argument

I don't understand why the code doesn't work when it did half an hour ago, nor why my computer thinks python is still using the pdfs, well after they've been merged (which happens almost instantaneously).

Thanks much for your help and apologies if these are stupid questions and/or badly formatted. I'm always looking to learn.

**Comment by MartinThoma** [comment]
The problem is that you `open` a file handle, but you don't close it.

instead of opening a file handle, just pass the path directly to PyPDF2:

```python
# Old and error-prone: Opening a file handle directly:
file_handle = open(filename, "rb")
reader = PyPDF2.PdfReader(file_handle)

# New: Let PyPDF2 manage file handles:
reader = PyPDF2.PdfReader(filename)
```

**Reply by DeanStra** [reply]
Thanks. I made the change you suggested and get 

> PdfReadError: Cannot read an empty file

Sorry again for being dense, what do I need to do? Just in case, the code I'm running is below:

```
import PyPDF2, os

# Get all the PDF filenames.
pdfFiles = []
for filename in os.listdir('.'):
    if filename.endswith('.pdf'):
        pdfFiles.append(filename)
pdfFiles.sort()

pdfWriter = PyPDF2.PdfFileWriter()

# Loop through all the PDF files.
for filename in pdfFiles:
    pdfReader = PyPDF2.PdfFileReader(filename)

    # Loop through all the pages (except the first) and add them.
    for pageNum in range(0, pdfReader.numPages):
        pageObj = pdfReader.getPage(pageNum)
        pdfWriter.addPage(pageObj)

# Save the resulting PDF to a file.
pdfOutput = open('allminutes.pdf', 'wb')
pdfWriter.write(pdfOutput)
pdfOutput.close()
```

**Reply by MartinThoma** [reply]
What does the error message say?
